src/indexing.jl

```diff
@@ -0,0 +1,395 @@
+    $error_if_canonical(::IndexStyle, A::AbstractArray, ::Any...) = nothing
+  end
+end
+
+# AbstractArrayInterface fallback definitions
+# -------------------------------------------
+@interface ::AbstractArrayInterface isstored(A::AbstractArray, i::Int, I::Int...) = (
+  @inline; @boundscheck checkbounds(A, i, I...); true
+)
+
+@interface ::AbstractArrayInterface function getunstoredindex(A::AbstractArray, I::Int...)
+  @inline
+  @boundscheck checkbounds(A, I...)
+  return zero(eltype(A))
+end
+@interface ::AbstractArrayInterface getstoredindex(A::AbstractArray, I::Int...) = (
+  @inline; getindex(A, I...)
+)
+@interface ::AbstractArrayInterface setstoredindex!(A::AbstractArray, v, I::Int...) = (
+  @inline; setindex!(A, v, I...)
+)
+@interface ::AbstractArrayInterface setunstoredindex!(A::AbstractArray, v, I::Int...) =
+  error("setunstoredindex! for $(typeof(A)) is not supported")
+
+@interface ::AbstractArrayInterface eachstoredindex(A::AbstractArray, B::AbstractArray...) =
+  eachstoredindex(IndexStyle(A, B...), A, B...)
+@interface ::AbstractArrayInterface eachstoredindex(
+  style::IndexStyle, A::AbstractArray, B::AbstractArray...
+) = eachindex(style, A, B...)
+
+@interface ::AbstractArrayInterface storedvalues(A::AbstractArray) = values(A)
+@interface ::AbstractArrayInterface storedpairs(A::AbstractArray) = pairs(A)
+@interface ::AbstractArrayInterface storedlength(A::AbstractArray) = length(storedvalues(A))
+
+# SparseArrayInterface implementations
+# ------------------------------------
+# canonical errors are moved to `isstored`, `getstoredindex` and `getunstoredindex`
+# so no errors at this level by defining both IndexLinear and IndexCartesian
+@interface ::AbstractSparseArrayInterface function Base.getindex(
+  A::AbstractArray{<:Any,N}, I::Vararg{Int,N}
+) where {N}
+  @_propagate_inbounds_meta
+  @boundscheck checkbounds(A, I...) # generally isstored requires bounds checking
+  return @inbounds isstored(A, I...) ? getstoredindex(A, I...) : getunstoredindex(A, I...)
+end
+@interface ::AbstractSparseArrayInterface function Base.getindex(A::AbstractArray, I::Int)
+  @_propagate_inbounds_meta
+  @boundscheck checkbounds(A, I)
+  return @inbounds isstored(A, I) ? getstoredindex(A, I) : getunstoredindex(A, I)
+end
+# disambiguate vectors
+@interface ::AbstractSparseArrayInterface function Base.getindex(A::AbstractVector, I::Int)
+  @_propagate_inbounds_meta
+  @boundscheck checkbounds(A, I)
+  return @inbounds isstored(A, I) ? getstoredindex(A, I) : getunstoredindex(A, I)
+end
+
+@interface ::AbstractSparseArrayInterface function Base.setindex!(
+  A::AbstractArray{<:Any,N}, v, I::Vararg{Int,N}
+) where {N}
+  @_propagate_inbounds_meta
+  @boundscheck checkbounds(A, I...)
+  return @inbounds if isstored(A, I...)
+    setstoredindex!(A, v, I...)
+  else
+    setunstoredindex!(A, v, I...)
+  end
+end
+@interface ::AbstractSparseArrayInterface function Base.setindex!(A::AbstractArray, I::Int)
```

Comment:
```suggestion
@interface ::AbstractSparseArrayInterface function Base.setindex!(A::AbstractArray, v, I::Int)
```
Looks like this also needs a test.

Reply:
Good catch, thanks!